### ISSUE TITLE
[mmlspark=0.17] LightGBMRegressor failing due to Connection Exception

### ISSUE THREAD

**Issue by skale1990** [Azure/mmlspark#646]
environment= spark 2.4 on k8s
Executor stacktrace
==============
`[2019-08-12T10:15:07.103Z] spark.org.apache.spark.executor.Executor ERROR: appId: 901C9278-B4E9-4EBB-AABF-8DB9C4DEF974, du: 1, message: Exception in task 1.0 in stage 30.0 (TID 1429)
java.lang.NullPointerException
	at com.microsoft.ml.spark.TrainUtils$.networkInit(TrainUtils.scala:275)
	at com.microsoft.ml.spark.TrainUtils$.trainLightGBM(TrainUtils.scala:313)
	at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
	at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:402)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:408)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[2019-08-12T10:15:07.103Z] spark.org.apache.spark.executor.Executor ERROR: appId: 901C9278-B4E9-4EBB-AABF-8DB9C4DEF974, du: 1, message: Exception in task 1.0 in stage 37.0 (TID 2034)
java.lang.NullPointerException
	at com.microsoft.ml.spark.TrainUtils$.networkInit(TrainUtils.scala:275)
	at com.microsoft.ml.spark.TrainUtils$.trainLightGBM(TrainUtils.scala:313)
	at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
	at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:402)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:408)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
[2019-08-12T10:15:07.427Z] spark.org.apache.spark.executor.Executor ERROR: appId: 901C9278-B4E9-4EBB-AABF-8DB9C4DEF974, du: 1, message: Exception in task 2.0 in stage 37.0 (TID 2037)
java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at java.net.Socket.<init>(Socket.java:434)
	at java.net.Socket.<init>(Socket.java:211)
	at com.microsoft.ml.spark.TrainUtils$.getNetworkInitNodes(TrainUtils.scala:240)
	at com.microsoft.ml.spark.TrainUtils$$anonfun$7.apply(TrainUtils.scala:303)
	at com.microsoft.ml.spark.TrainUtils$$anonfun$7.apply(TrainUtils.scala:298)
	at com.microsoft.ml.spark.StreamUtilities$.using(StreamUtilities.scala:29)
	at com.microsoft.ml.spark.TrainUtils$.trainLightGBM(TrainUtils.scala:297)
	at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
	at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:402)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:408)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)`

**Comment by skale1990** [comment]
[executor_log.zip](https://github.com/Azure/mmlspark/files/3492170/executor_log.zip)

Executor logs for reference.

**Comment by skale1990** [comment]
I'm running training code for 5 different groups in 5 parallel threads. so there are 5 LightGBMRegressor running at a time

**Comment by imatiach-msft** [comment]
@skale1990 sorry about the issue you are encountering.  I looked through the logs and I see some empty partitions:
```LightGBM worker encountered empty partition, for best performance ensure no partitions empty```
Also, I saw some null pointer exceptions on network init.
Do you still see errors with the new v0.18.1?  Also, do you have dynamic allocation disabled on the cluster?  To get rid of empty partitions, you could try to shuffle your dataset.

**Comment by jaipal1995** [comment]
Hi @imatiach-msft,
I have been facing the same issue when running for n different groups in n different threads. Could you please explain the cause?

**Comment by pmaloo** [comment]
Hi @imatiach-msft, I ran into the same issue, any updates on how to fix this?

**Comment by ranadeepsingh** [comment]
Thank you for the detailed executor evidence. This overlaps the LightGBM network family, but empty partitions and dynamic allocation make it a mixed-cause case rather than a proven duplicate of #1967. Keeping it separate pending a current-release reproduction.
